File: Model Selection/k-fold cross validation/k_fold_cross_validation.R
# Model Selection --> k - fold cross validation
# Here we evluate the model performance and improve the model performance. This inclues choosing the best parameters of the machine learning model.
# Everytime we build a machine learning model we have two types of parameters.
# [1] The parameer the model learnt. That is the parametes taht were changed and found the optimal values by running the model.
# [2] The parameters that we choose ourselves. These parameters are called the hyper parameters.Thus there is way of choosing the optimal values for these hyper parameters. This can be done through the grid search.
# Upto now we have the training set and test set. We train the data on the training set and we test the performance of the model on the test set.
# But doing this we have the variance problem. The variance problem can be explained by when we get the accuracy on the test set and test the model on a different test set we get a different accuracy.
# So judging our model on only one accuracy and one test set is not super relevant.
# This can be solved by k-fold cross validation. This can is done by splitting the training set into 10 folds when k = 10.
# Then we train the model on 9 folds and test the model on last fold.
# Since with 10 folds we can make 10 different combinations of 9 folds and 1 fold to test it. That is we can train and test the model on 10 different combinations.
# Hence we can take the average of the accuracies of the different 10 observations and also compute the standard deviation to look at the variance.
# The bias - variance tradeoff categories
# [1] Good accuracy and small variance --> Low bias low variance.
# [2] Large accuracy and high variance --> Low bias high variance.
# [3] Small accuracy and low variance --> High bias low variance.
# [4] Low accuaracy and high variance --> High bias high variance.

# Importing the dataset
dataset = read.csv('Social_Network_Ads.csv')
dataset = dataset[3:5]

# Encoding the target feature as factor
dataset$Purchased = factor(dataset$Purchased, levels = c(0, 1))

# Splitting the dataset into the Training set and Test set
install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(dataset$Purchased, SplitRatio = 0.75)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Feature Scaling
training_set[-3] = scale(training_set[-3])
test_set[-3] = scale(test_set[-3])

# Fitting classifier to the Training set
install.packages('e1071')
library(e1071)
classifier = svm(formula = Purchased ~ ., data = training_set, type = 'C-classification', kernel = 'radial')

# Predicting the Test set results
y_pred = predict(classifier, newdata = test_set[-3])

# Making the Confusion Matrix
cm = table(test_set[, 3], y_pred)

# Applying the K - Fold Cross Validation
install.packages('caret')
library(caret)
folds = createFolds(training_set$Purchased, k = 10)
# lapply function --> To apply the function to the different elements of the list.
# Here function --> Function used to find the accuracy of each of these 10 folds.
# list --> folds.
cv = lapply(folds, function(x){
  training_fold = training_set[-x, ]
  test_fold = training_set[x, ]
  classifier = svm(formula = Purchased ~ ., data = training_fold, type = 'C-classification', kernel = 'radial')
  y_pred = predict(classifier, newdata = test_fold[-3])
  cm = table(test_fold[, 3], y_pred)
  accuracy = (cm[1,1] + cm[2,2]) / (cm[1,1] + cm[2,2] + cm[2,1] + cm[1,2])
  return(accuracy)
})
accuracy = mean(as.numeric(cv))
# x --> Each one of the 10 folds.
# Inthis function we are going to implement k - fold cross validation. This contains:
# [1] training_set --> This is the whole training set from which we withdraw the test fold.

# Visualising the Training set results
library(ElemStatLearn)
set = training_set
X1 = seq(min(set[, 1]) - 1, max(set[, 1]) + 1, by = 0.01)
X2 = seq(min(set[, 2]) - 1, max(set[, 2]) + 1, by = 0.01)
grid_set = expand.grid(X1, X2)
colnames(grid_set) = c('Age', 'EstimatedSalary')
y_grid = predict(classifier, newdata = grid_set)
plot(set[, -3],
     main = 'Kernel SVM (Training set)',
     xlab = 'Age', ylab = 'Estimated Salary',
     xlim = range(X1), ylim = range(X2))
contour(X1, X2, matrix(as.numeric(y_grid), length(X1), length(X2)), add = TRUE)
points(grid_set, pch = '.', col = ifelse(y_grid == 1, 'springgreen3', 'tomato'))
points(set, pch = 21, bg = ifelse(set[, 3] == 1, 'green4', 'red3'))

# Visualising the Test set results
library(ElemStatLearn)
set = test_set
X1 = seq(min(set[, 1]) - 1, max(set[, 1]) + 1, by = 0.01)
X2 = seq(min(set[, 2]) - 1, max(set[, 2]) + 1, by = 0.01)
grid_set = expand.grid(X1, X2)
colnames(grid_set) = c('Age', 'EstimatedSalary')
y_grid = predict(classifier, newdata = grid_set)
plot(set[, -3], main = 'Kernel SVM (Test set)',
     xlab = 'Age', ylab = 'Estimated Salary',
     xlim = range(X1), ylim = range(X2))
contour(X1, X2, matrix(as.numeric(y_grid), length(X1), length(X2)), add = TRUE)
points(grid_set, pch = '.', col = ifelse(y_grid == 1, 'springgreen3', 'tomato'))
points(set, pch = 21, bg = ifelse(set[, 3] == 1, 'green4', 'red3'))

File: Model Selection/k-fold cross validation/k_fold_cross_validation.py
# K-fold cross validation --> Optimizing way to evluate the models.
# Here we evaluate the model performance and improve the model performance.
# That is choosing the best parameters of the machine learning model.
# There are two types of parameters
# [1] --> Parameters that the model learns and the parameters that were changed and found the optimal values by running the model.
# [2] --> The parameters that we choose ourselves.
# These parameters are called the hybrid parameters.
# In previous models we split the data in to training set and the test set. Then we train the data on the training set and test the data in the test set. But here we get the variance problem. This can be explained by when we get the accuracy on the test set. So if we run the model and test the model on the other test set we would get another accuracy.
# So judjing our model only on one accuracy on one test set is not super relevant. This is not the mot relevant way to test the model performance.
# The k-fold cross validation will fix this problem. This is fixed by spinning the training set in to ten folds when k = 10.
# Then we train our model on 9 folds and test is on the remaining fold.
# Thus with the 10 folds we get 9 combinations of 9 folds to train the model and 1 fold to test the model.
# Thus we can train and test the model on 10 sets of combinations.
# Then we take the accuracy of different 10 evaluations and also compute the standard deviation to see the variance.
# Bias-Variance Trade off
# [1] Good accuracy and small variance --> Low Bias Low Variance
# [2] Large accuaracy and high variance --> Low Bias High Variance
# [3] Small accuracy and low variance --> High Bias Low Variance
# [4] Low accuaracy and high variance --> High Bias High Variance

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Social_Network_Ads.csv')
X = dataset.iloc[:, [2,3]].values
y = dataset.iloc[:, 4].values

# Splitting the dataset into the Training set and Test set
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0)

# Feature Scaling
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)

# Fitting Classifier to Training Set
from sklearn.svm import SVC
classifier = SVC(kernel = 'rbf', random_state = 0)
classifier.fit(X_train, y_train)

# Predicting the Test Set Results
# y_pred --> Vector of predictions
y_pred = classifier.predict(X_test)

# Making the Confusion Matrix
# Confusion Matrix -> To see weather our Logistic Regression made correct prediction or not.
# This confusion matrx will contain the correct predictions made on the Test Set as well as the incorrect predictions,
# For this we are importing a function and not a class.
# Distinction --> Class contains the captial letters at the beginning.
# Parameters of cnfusion matrix -> (1) y_true = Real values thats the values of the data set, (2) y_pred.
# 65, 24 = 89-> Correct Predictions, 8,3 = 11-> Incorrect Predictions
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)

# Applying k - fold cross validation.
from sklearn.model_selection import cross_val_score
accuracies = cross_val_score(estimator = classifier, X = X_train, y = y_train, cv = 10)
accuracies.mean()
accuracies.std()
# Standard Deviation --> That is the average of the different accuracies that we will get when evluating the model performance and the average accuracy that is 90 percentage is 6 percentage. That means we are in low bias and low variance category. That is not too high variance.

# Parameters
# [1] estimator --> classifier.
# [2] X --> training set.
# [3] y --> Dependent variable vector of the training set.
# [4] cv --> Number of folds we want to split the training set into.
# [5] n_jobs --> -1 (Set this to -1 if we have the larger dataset. This will enable all the CPUs in the computer.)

# Visualizing the Training Set Results
from matplotlib.colors import ListedColormap
X_set, y_set = X_train, y_train
X1, X2 = np.meshgrid(np.arange(start = X_set[:, 0].min() - 1, stop = X_set[:, 0].max() + 1, step = 0.01),
                     np.arange(start = X_set[:, 1].min() - 1, stop = X_set[:, 1].max() + 1, step = 0.01))
plt.contourf(X1, X2, classifier.predict(np.array([X1.ravel(), X2.ravel()]).T).reshape(X1.shape),
             alpha = 0.75, cmap = ListedColormap(('red', 'green')))
plt.xlim(X1.min(), X1.max())
plt.ylim(X2.min(), X2.max())
for i, j in enumerate(np.unique(y_set)):
    plt.scatter(X_set[y_set == j, 0], X_set[y_set == j, 1],
                c = ListedColormap(('red', 'green'))(i), label = j)
plt.title('Logistic Regression (Training set)')
plt.xlabel('Age')
plt.ylabel('Estimated Salary')
plt.legend()
plt.show()

# Visualising the Test set results
from matplotlib.colors import ListedColormap
X_set, y_set = X_test, y_test
X1, X2 = np.meshgrid(np.arange(start = X_set[:, 0].min() - 1, stop = X_set[:, 0].max() + 1, step = 0.01),
                     np.arange(start = X_set[:, 1].min() - 1, stop = X_set[:, 1].max() + 1, step = 0.01))
plt.contourf(X1, X2, classifier.predict(np.array([X1.ravel(), X2.ravel()]).T).reshape(X1.shape),
             alpha = 0.75, cmap = ListedColormap(('red', 'green')))
plt.xlim(X1.min(), X1.max())
plt.ylim(X2.min(), X2.max())
for i, j in enumerate(np.unique(y_set)):
    plt.scatter(X_set[y_set == j, 0], X_set[y_set == j, 1],
                c = ListedColormap(('red', 'green'))(i), label = j)
plt.title('Logistic Regression (Test set)')
plt.xlabel('Age')
plt.ylabel('Estimated Salary')
plt.legend()
plt.show()